### ISSUE TITLE
Prefix matching for rating subject

### ISSUE THREAD

**Issue by andmar** [cgrates/cgrates#388]
Hy.

Here is a feature request for "Prefix matching for rating subject". The main driver of this request is to allow calls with different CLI + Destination combinations (both with prefix matching) to be billed/LCRed differently by CGRates. You may also hear/see this feature refered to as A-number/CLI billing/pricing and as we understand is becoming an issue for carriers in Europe.

Thanks for the effort @danbogos, looking forward to hear from you guys about this. 

Best regards.

**Comment by KeithMilner** [comment]
This sounds similar to the situation I was dealing with for US carrier rates:
https://groups.google.com/forum/#!topic/cgrates/Gzc4szHHLGY

**Comment by rif** [comment]
For enabling prefix matching  you have to set rp_subject_prefix_matching option in rater config section to true.

https://github.com/cgrates/cgrates/blob/master/data/conf/cgrates/cgrates.json#L80

Please test and report any issues.

**Comment by MarcVoIP** [comment]
Rif, can you describe briefly how this feature work ? Is it suitable also for inter/intra-state billing in US ?

**Comment by rif** [comment]
When searching for a matching rating profile it will search the entire subject first then it will try one character less till either match or only one subject character is left.
If not found the fallback subjects are processed with the same procedure.

I am not sure how the inter/intra-state billing in US work, does the above sound useful for that?

**Comment by MarcVoIP** [comment]
I see, so this is actually related to the matching digits count.
Inter/intra-state routing is based on different rules and main issue is the destination rate depends from where the call was origniated (initial digits from the CLI) and where the call is terminated (initial digits from the destination number). 
More details : https://groups.google.com/forum/#!topic/cgrates/Gzc4szHHLGY

@KeithMilner : this will not help with the US termination case however this feature is definitely useful in general.

**Comment by cgrates** [comment]
@MarcVoIP, I think this feature relates exactly what you have described: 
the ability to match both CLI and Destination based on prefix when 
finding the rating profile.

DanB

**Comment by MarcVoIP** [comment]
Dan, I guess the example of use would help us understand how it works then. 
An example of inter/instra-state billing is described on the group discussion.

**Comment by danbogos** [comment]
@MarcVoIP,

Example of use:

Call with:
{
Subject: +4986517174963,
Destination: +11231234
}

RatingProfile definition:
*out,cgrates.org,call,*any,2014-01-14T00:00:00Z,RP_RETAIL1,,
*out,cgrates.org,call,+49,2014-01-14T00:00:00Z,RP_RETAIL2,,


With rp_subject_prefix_matching = false we will match RETAIL1 since the 
rating subject searched will be +4986517174963 then fallback on *any.
With rp_subject_prefix_matching = true we will match RP_RETAIL2 since 
the rating subject searched will be +4986517174963, +498651717496, 
+49865171749, +4986517174, +498651717, +49865171, +4986517, +498651, 
+49865, +4986, +498, +49 with match on last one.

Before rp_subject_prefix_matching we have had only the ability to match 
on destination prefix but full CLI. Now we can also match on CLI prefix 
(which as per my understanding is what you need to define your 
intra-state rules, eg: if the CLI starts with +123 means intrastate if 
starts with +122 means in-state). The downside of this is a bit of more 
CPU power used (although the complete info is in cache).

Hope it is a bit more clear now.

DanB

**Comment by MarcVoIP** [comment]
Thank you Dan for the detailed instruction. This makes things clear. 

Unfortunately I still don't see the ability to do inter/intra-state billing and routing in this case. 

Inter/Intra billing flow:
 There are about 180k NPA/NXX codes in US. Each phone number starts with country code (1) and then NPANXX. The list of codes is static of course so it can be easily obtained.
  Example:    1646251xxxx is a New York number (NY)    and 1772234 is from Florida (FL) 
 
These two numbers are from different states so if NY calls FL or FL calls NY then the call will be considered as interstate (originated from one state and terminated in another state. 

The decision whether the call is inter or intra-state is made based on the lookup to NPANXX table for two numbers (ANI and destination). 
Then, once we know which type of call we're dealing with the appropriate rate should be selected to bill the call. 

Another level of complication is the fact that US carriers charge differently for inter and intrastate calls  so this should be also considered in the LCR engine to make a decision which carrier should be used first based on given type of call (inter or intra call).

This would make no sense to add all possible combinations to the rating profiles thus the feature "Prefix matching for rating subject " is not resolving that issue. 

I would love to see good solution for inter/intra calls within CGRateS. So far I'm using some kind of workaround by adding special prefix before the destination number before the call actually hits the billing so my rate tables are doubled and each "npanxx" is added twice: one entry for interNPANXX and one entry for intraNPANXX. So far this works for me.

**Comment by danbogos** [comment]
MarcVoIP,

What about aliasing with Subject prefix matching?

EG: 1646251xxxx will be aliased to "NY" then in your rating profile you 
will have a line with costs for NY calling different destinations 
(including NY or or FL since destinations can be also aliased).

Will that work?

DanB

**Comment by MarcVoIP** [comment]
I'm trying to understand your suggested solution but not really sure how this suppose to work.
In any way - would that be still suitable for 180k different NPANXX codes ? There must be a rule for each and every combination in the system as any NPANXX can call any other NPANXX so based on that we figure out if the call is inter or intra.

**Comment by danbogos** [comment]
I would say yes. We have cases when it works for millions of 
destinations without any issue.

DanB

**Comment by andmar** [comment]
Will this apply to LcrRules also or that would require aditional work?

**Comment by KeithMilner** [comment]
I will point out that any NPANXX to any other NPANXX is over 32 BILLION potential combinations.

However, I will also point out that, whilst there are 180k NPANXX codes, there are only 50 states, and the selection rule for which set of rates to use is based on CgPN state = CdPN ( not on CgPN NPANXX = CdPN NPANXX). So we just need to reduce the matching to state level.

TBH I understand the aliasing capabilities well enough to know if this is possible or not within that system.

The approach I've been taking, to date, is similar to MarcVoIP's with two tables (and prefixes) and computing the prefix outside of CGRateS.

In my case, I have an mtree in kamailio containing the NPA codes and states codes. This is used by some logic within kamailio to look up the CgPN and CdPN state codes and compare them. This is used to add the appropriate prefix.

If it helps I'm happy to share this code so you can understand the mechanism. That is, assuming, we are keen to get this working wholly within CGRateS.

**Comment by cgrates** [comment]
Hi Keith,

Thanks for your input. Unfortunately we need a more global solution than
specific kamailio implementation.

Regarding Aliasing, what I was suggesting was completely renaming rating
subject to reflect the exact logic you want: a region calls another region
since you don't care who is the originator. If you do care to say: CLI A
calling from location B calling destination C then you will need to alias
category instead of subject. However that is more or less trivial with
aliases in the same way as I have proposed.

DanB

**Comment by andmar** [comment]
@danbogos / @rif will this apply to LcrRules also or that would require aditional work? I just remembered that when using LCRs the matching on the Subject for the call would be defined there instead of the Routing Profile...

**Comment by rif** [comment]
Yes, this applies to LcrRules too.

**Comment by andmar** [comment]
Cool. Thanks a lot for this feature. :+1:

**Comment by cgrates** [comment]
Hey Andre,

Just to be clear: the prefix matching works for finding the prices for LCR
(RatingProfiles.csv) but not for filtering in LcrProfiles.csv.

Hope this helps.

DanB

**Comment by andmar** [comment]
Hy Dan.

So, that now got me a little confused.

Imagine we need to apply the following:

> Calls with CLI (subject) +351 and destination +34 get one LCR.
> Calls with CLI (subject) +244 and destination +34 get another LCR.

How would you see it be done? I was going to try this by inserting two lines in the LcrRules file and then the system would match by prefix there and associate them with an lcr_profile definition in the RoutingProfiles file.

If this is not possible is the alternative configuring the RPCategory with the "CLI" information on the LcrRules?

**Comment by danbogos** [comment]
Hi Andre,

We will need prefix matching also in lcrRules for that to work. If that 
is the case, please open another issue and we should check it.

DanB

**Comment by KeithMilner** [comment]
I have to admit I'm still a bit lost on how aliases can be used for this.

I understand that aliasing can be used to transform the subject (containing the CLI) and the destination to reduce them down to states, but this is only useful for selecting which subset of destination rates should be used. This reduced subject and destination can't then be used to determine the specific rate to use.

Some examples:
If CLI=319-200-1234 called Destination=563-945-9876, these could alias down to subject=IA destination=IA
Another call, CLI=319-200-1234 called Destination=641-204-5555, this would also alias down to subject=IA destination=IA

But both these calls have different rates, and "IA calls IA" isn't enough to determine which rate. It's only enough to determine which set of rates to use. The full original destination number or at least the first 6 digits of it) are then required (as a subsequent step) to find the rate.

Keith

**Comment by andmar** [comment]
Hy Dan.

Ok, I'll do that, thanks. :) For us is precious, but also, I think it's interesting for you guys to support the feature also in the LCRs.

**Comment by danbogos** [comment]
Hi Keith,

What I was thinking is that you replace category part with the one 
matching your from state (eg: call_from_IA). Then you can apply rates to 
calls from IA (including selection per subject).
Does that work?

DanB

**Comment by KeithMilner** [comment]
Hi Dan,
It might be possible to make that work, albeit in a fairly clunky way which, TBH, makes doing the lookup in Kamailio/Freeswitch still a very attractive option:

If the category was call_from_<STATE> then this would, presumably, be used to drive a <STATE> specific RatingPlan. Each State's rating plan would need to have DestinationRates entries for every NPANXX (all 180k+ of them), with the same-State entries coming from the Intrastate rates, and the other-State entries coming from the Interstate rates.

This makes the number of rate entries a lot lower than having to deal with every NPANXX combination individually (the 32 billion option) but it's still 9 million DestinationRates (and associated Rates) entries just for mainland USA, with a huge amount of duplication. I'm sure CGRateS can cope with this, but it has to impact performance somehow.

Possibly more importantly, it's a lot of data to manage and I can foresee some challenges getting the data into this format without introducing errors, performing validation, etc. every time the rates change.

Compare this with the approach I'm currently taking (computing the States and the Inter/Intrastate in Kamailio) which only needs the 180k table as-is which, as it has two rate columns, is 360k DestinationRates, copied directly from the carrier rate tables (plus associated Rates). That's a lot easier to manage.

So, yes, this is a solution, and I suspect it can be made to work, but it's not a particularly attractive one (to me at least) given there is a more efficient and easier to manage alternative.

Cheers,

Keith

**Comment by danbogos** [comment]
Keith,

OK, nothing against your approach if you are happy with it.
We did not have many queries on intra/inter-state for US but as soon as 
the approach proposed will become an issue we can think about optimizing it.

DanB

**Comment by KeithMilner** [comment]
I'm happy as is. We have a solution that, whilst not 100% optimal (in that part of the solution is managed outside CGRateS) it's easy to manage and efficient.

I've published my implementation of this using kamailio onto the Google Groups thread so it can be used by others and it should be relatively easy to re-implement this on FreeSwitch, Asterisk, etc.

In any case, in the kamailio implementation I have published I've not specified what the interface with CGRateS is. There may be cases where it makes sense to push this information to CGRateS as part of the subject, versus others where it makes sense to send it as a prefix to the destination, etc.
